codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -378,16 +384,34 @@ protected Http2FrameWriter frameWriter() {
      * Writes a {@code RST_STREAM} frame to the remote endpoint and updates the connection state appropriately.
      */
     @Override
-    public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
-            ChannelPromise promise) {
-        Http2Stream stream = connection().stream(streamId);
+    public ChannelFuture writeRstStream(final ChannelHandlerContext ctx, int streamId, long errorCode,
+            final ChannelPromise promise) {
+        final Http2Stream stream = connection().stream(streamId);
+        if (stream == null) {
+            // Don't write a RST_STREAM frame if we are not aware of the stream.
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
```

Comment:
@Scottmitch `stream == null` could indicate that the stream never existed or that it's closed. So treating RST_STREAM for streams in the closed as an error is again problematic.